src/main/kotlin/org/cascadebot/cascadebot/utils/placeholders/Placeholders.kt

```diff
@@ -0,0 +1,164 @@
+import org.cascadebot.cascadebot.data.language.Locale
+import org.cascadebot.cascadebot.utils.FormatUtils
+import java.time.Duration
+import java.time.OffsetDateTime
+
+class Placeholders<T>(var key: String) {
+
+    companion object {
+        // https://regex101.com/r/LxMI0R/3
+        val placeholderRegex = Regex("\\{([\\p{Ll}_-]+)(?::((?:[\\p{Ll}0-9_-]+,?)+))?}")
+    }
+
+    private val _placeholders: MutableList<Placeholder<T>> = mutableListOf()
+    val placeholders: List<Placeholder<T>>
+        get() = _placeholders.toList()
+
+    fun staticPlaceholder(key: String, mapping: StaticPlaceholder<T>.(PlaceholderContext<T>) -> String?) {
+        _placeholders.add(StaticPlaceholder(key, "${this.key}.$key", mapping))
+    }
+
+    fun argsPlaceholder(key: String, mapping: ArgsPlaceholder<T>.(PlaceholderContext<T>, List<String>) -> String?) {
+        _placeholders.add(ArgsPlaceholder(key,"${this.key}.$key", mapping))
+    }
+
+    // TODO: Use locale!
+    fun formatMessage(locale: Locale, message: String, input: T): String {
+        val toReplace = mutableMapOf<String, String>()
+        for (matchResult in placeholderRegex.findAll(message)) {
+            _placeholders.find { it.localisedInfo.values.any { info -> info.key == matchResult.groupValues[1] } }?.let { placeholder ->
+                when (placeholder) {
+                    is StaticPlaceholder<T> -> {
+                        placeholder.mapping(placeholder, PlaceholderContext(input, locale))?.let {
+                            toReplace[matchResult.groupValues[0]] = it
+                        }
+                    }
+                    is ArgsPlaceholder<T> -> {
+                        val args = if (matchResult.groupValues[2].isNotEmpty()) {
+                            matchResult.groupValues[2].split(",")
+                        } else {
+                            listOf()
+                        }
+
+                        placeholder.mapping(placeholder, PlaceholderContext(input, locale), args)?.let {
+                            toReplace[matchResult.groupValues[0]] = it
+                        }
+                    }
+                    else -> error("Invalid Placeholder type: ${placeholder::class.simpleName}")
+                }
+            }
+        }
+        var newMessage = message
+        toReplace.forEach { newMessage = newMessage.replace(it.key, it.value) }
+        return newMessage
+    }
+
+    fun highlightMessage(message: String): String {
+        val toReplace = mutableMapOf<String, String>()
+        for (matchResult in placeholderRegex.findAll(message)) {
+            val placeholder = _placeholders.find { it.localisedInfo.values.any { info -> info.key == matchResult.groupValues[1] } }
+            if (placeholder != null) {
+                matchResult.groupValues[0].also { toReplace[it] = "`$it`" }
+            } else {
+                matchResult.groupValues[0].also { toReplace[it] = "~~$it~~" }
```

Comment:
Someone could easily accidentally break this formatting. You might want to strip existing formatting

Reply:
Not needed since `{placeholders}` cannot contain symbols